### PR TITLE
Initialize the whole BitArray by data value

### PR DESCRIPTION
**Pull request by 2henwei**
If `array_size` is greater than `data_size`, only the first `data_size` bytes will be initialized. In a zero-initialized scenario, the reset of BitArray are not correctly initialized. It would be nice to have it and doesn't influence the original functionality.

### CODE REVIEW

**Comment by 2henwei** [comment]
@knute-sifive I tried to reproduce the bug on `Register_test` but I can't.

Then I found the problem may be caused by the following code snippet:
https://github.com/sparcians/map/blob/c9c12b24dec0c54ed0decd62a810d1ace7766c34/sparta/sparta/utils/BitArray.hpp#L221-L225

```data_ss_``` was initialized to zero at the beginning so the ```write_mask``` is also correctly initialized to zero. In `Register_test`, large register is assigned exactly 16-bytes, which fits ```SMALL_OPTIMIZATION_SIZE```.

https://github.com/sparcians/map/blob/c9c12b24dec0c54ed0decd62a810d1ace7766c34/sparta/sparta/functional/Register.hpp#L1327-L1333

On the other hand, register size exceeding 16-bytes will be allocated an non-initialized memory space, ```data_ls_```. Its ```write_mask``` may hence have some issues. So I suggested:

1. Initialize the ```data_ls_``` after the memory space is allocated. (So the original commit in this PR doesn't need anymore)
2. Add a "huge" register test (32-bytes) in Register_test.

**Comment by knute-sifive** [comment]
> So I suggested:
> 
>     1. Initialize the `data_ls_` after the memory space is allocated. (So the original commit in this PR doesn't need anymore)
> 
>     2. Add a "huge" register test (32-bytes) in Register_test.
> 
> 
> What do you think?

Sounds like a good plan!

**Comment by 2henwei** [comment]
@knute-sifive, I add a simple huge (256b) register test (basically the same as the large register test). If you directly checkout that commit 7126c5ce8eba9f9fa558014e2f07db7b77045ea3, you would find the following error:

```
Writing to <top.dummy.regs.huge 256 bits cc cc cc cc cc cc cc cc cc cc cc cc cc cc cc cc cc cc 
cc cc cc cc cc cc cc cc cc cc cc cc cc cc>                                                     
 now: <top.dummy.regs.huge 256 bits cc cc cc cc cc cc cc cc cc cc cc cc cc cc cc cc cc cc cc cc
 cc cc cc cc ff ff ee fd ff ff ff ff>                                                          
 now: <top.dummy.regs.huge 256 bits cc cc cc cc cc cc cc cc cc cc cc cc cc cc cc cc cc cc cc cc
 cc cc cc cc ee ee ee ec ff ff ff ff>                                                          
 now: <top.dummy.regs.huge 256 bits cc cc cc cc cc cc cc cc cc cc cc cc cc cc cc cc cc cc cc cc
 cc cc cc cc dd dd ee ec ff ff ff ff>                                                          
 now: <top.dummy.regs.huge 256 bits cc cc cc cc cc cc cc cc cc cc cc cc cc cc cc cc cc cc cc cc
 cc cc cc cc dd dd ee ec dd dd ff ff>                                                          
 now: <top.dummy.regs.huge 256 bits cc cc cc cc cc cc cc cc cc cc cc cc cc cc cc cc cc cc cc cc
 cc cc cc cc cc dd ee ec dd dd ff ff>                                                          
 now: <top.dummy.regs.huge 256 bits cc cc cc cc cc cc cc cc cc cc cc cc cc cc cc cc cc cc cc cc
 cc cc cc cc cc dd ee ec cc dd ff ff>                                                          
Test '(huge->read<uint64_t>())' FAILED on line 956 in file /home/zhenw/athena-devel/map/sparta/
test/Register/Register_test.cpp. Value: '14762217854407147519' should equal '147622178629970821
11'                                                                                            
```

File: sparta/test/Register/Register_test.cpp [hunk 1]
```diff
@@ -929,6 +937,25 @@ int main()
...
+    std::cout << " now: " << huge << std::endl;
+    EXPECT_NOTHROW((huge->write<uint16_t>(0xdddd, 1))); // 3-2
+    std::cout << " now: " << huge << std::endl;
+    EXPECT_NOTHROW((huge->write<uint8_t>(0xcc, 7))); // 7-0 (MSB)
+    std::cout << " now: " << huge << std::endl;
+    EXPECT_NOTHROW((huge->write<uint8_t>(0xcc, 3))); // 7-4
+    std::cout << " now: " << huge << std::endl;
+    std::cout << " have: " << std::hex << huge->read<uint64_t>() << " expect: " << 0xccddeeeeccddffff << std::endl;
+    EXPECT_EQUAL((huge->read<uint64_t>()), 0xccddeeeeccddffff);
+    std::cout << "Huge Register: " << std::endl << huge->renderSubtree(-1, true);
```
Comment (knute-sifive) [inline sparta/test/Register/Register_test.cpp:957]: This testing is great.. a few changes requested:
1. After each write to the given index, add an `EXPECT_EQUAL` to ensure that index was written properly:
```
EXPECT_NOTHROW((huge->write<uint32_t>(0xeeeeeeee, 1))); // 1-0 (MSB)
EXPECT_EQUAL(huge->read<uint32_t>(1). 0xeeeeeeee);
```
1. The `cout` statements aren't too useful if no one looks at them.  :)  Turn them into `EXPECT_EQUAL` statements.

File: sparta/sparta/utils/BitArray.hpp [hunk 2]
```diff
@@ -220,6 +220,7 @@ namespace sparta {
 
                 if (data_size > SMALL_OPTIMIZATION_SIZE) {
                     data_ls_.reserve(data_size);
+                    std::memset(&data_ls_[0], 0, data_size * sizeof(data_type));
```
Comment (knute-sifive) [inline sparta/sparta/utils/BitArray.hpp:223]: Instead of reserving the memory, resize the vector with the value 0.

Comment (2henwei) [reply sparta/sparta/utils/BitArray.hpp]: How could I forget this function? Thanks!

File: sparta/test/Register/Register_test.cpp (hunk 1 above)
Comment (2henwei) [reply sparta/test/Register/Register_test.cpp]: Done.

**Comment by 2henwei** [comment]
@knute-sifive Thanks for your reviewing and approval! Just a kind reminder that I don't have write access to this repository. (I cant merge it)

**Pull request merged**